SU2_CFD/src/output/filewriter/CParallelDataSorter.cpp

```diff
@@ -0,0 +1,412 @@
+      dataBuffer[ii*VARS_PER_POINT+jj] = tmpBuffer[ii];
+    }
+  }
+  
+  delete [] tmpBuffer;
+  
+  /*--- Store the total number of local points my rank has for
+   the current section after completing the communications. ---*/
+  
+  nParallel_Poin = nPoint_Recv[size];
+  
+  /*--- Reduce the total number of points we will write in the output files. ---*/
+
+#ifndef HAVE_MPI
+  nGlobal_Poin_Par = nParallel_Poin;
+#else
+  SU2_MPI::Allreduce(&nParallel_Poin, &nGlobal_Poin_Par, 1,
+                     MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+#endif
+  
+  /*--- Free temporary memory from communications ---*/
+  
+  delete [] idRecv;
+}
+
+void CParallelDataSorter::PrepareSendBuffers(std::vector<unsigned long>& globalID){
+  
+  unsigned long iPoint;
+  unsigned short iProcessor;
+  
+  int VARS_PER_POINT = GlobalField_Counter;
+  
+  /*--- We start with the grid nodes distributed across all procs with
+   no particular ordering assumed. We need to loop through our local partition
+   and decide how many nodes we must send to each other rank in order to
+   have all nodes sorted according to a linear partitioning of the grid
+   nodes, i.e., rank 0 holds the first ~ nGlobalPoint()/nProcessors nodes.
+   First, initialize a counter and flag. ---*/
+  
+  for (iPoint = 0; iPoint < nLocalPoint_Sort; iPoint++ ) {
+    
+    iProcessor = linearPartitioner->GetRankContainingIndex(globalID[iPoint]);      
+    
+    /*--- If we have not visited this node yet, increment our
+       number of elements that must be sent to a particular proc. ---*/
+    
+    nPoint_Send[iProcessor+1]++;
+  }
+  
+  /*--- Communicate the number of nodes to be sent/recv'd amongst
+   all processors. After this communication, each proc knows how
+   many cells it will receive from each other processor. ---*/
+  
+#ifdef HAVE_MPI
+  SU2_MPI::Alltoall(&(nPoint_Send[1]), 1, MPI_INT,
+                    &(nPoint_Recv[1]), 1, MPI_INT, MPI_COMM_WORLD);
+#else
+  nPoint_Recv[1] = nPoint_Send[1];
+#endif
+  
+  /*--- Prepare to send coordinates. First check how many
+   messages we will be sending and receiving. Here we also put
+   the counters into cumulative storage format to make the
+   communications simpler. ---*/
+  
+  nSends = 0; nRecvs = 0;
+
+  for (int ii = 0; ii < size; ii++) {
+    if ((ii != rank) && (nPoint_Send[ii+1] > 0)) nSends++;
+    if ((ii != rank) && (nPoint_Recv[ii+1] > 0)) nRecvs++;
+    
+    nPoint_Send[ii+1] += nPoint_Send[ii];
+    nPoint_Recv[ii+1] += nPoint_Recv[ii];
+  }
+  
+  /*--- Allocate memory to hold the connectivity that we are
+   sending. ---*/
+  
+  connSend = NULL;
+  connSend = new su2double[VARS_PER_POINT*nPoint_Send[size]]();
+  
+  
+  /*--- Allocate the data buffer to hold the sorted data.
+   * It should be big enough to also temporarily hold the data received using the AD type ---*/
+  
+  dataBuffer = reinterpret_cast<passivedouble*>(new su2double[VARS_PER_POINT*nPoint_Recv[size]]());
```

Comment:
this is messed up

Reply:
I want to avoid allocating additional memory, so I reuse the same memory using two pointers (one for su2double datatype and one for passivedouble). I changed it a little bit so that it is more obvious and that there will be now misleading new/delete.